### PR TITLE
GH-100719: Remove redundant `gi_code` field from generator object.

### PR DESCRIPTION
**Pull request by markshannon**
* Issue: gh-100719

### CODE REVIEW

File: Python/ceval.c [hunk 1]
```diff
@@ -1962,44 +1962,6 @@ initialize_locals(PyThreadState *tstate, PyFunctionObject *func,
     return -1;
 }
 
-/* Consumes references to func, locals and all the args */
```
Comment (Fidget-Spinner) [inline Python/ceval.c:1607]: Why was this moved downwards?

Comment (markshannon) [reply Python/ceval.c:1607]: There was a small change, it now calls `clear_thread_frame()` rather than `_PyEvalFrameClearAndPop()`, so now needs to come after `clear_thread_frame`

File: Include/cpython/genobject.h [hunk 2]
```diff
@@ -46,6 +44,7 @@ PyAPI_FUNC(PyObject *) PyGen_NewWithQualName(PyFrameObject *,
 PyAPI_FUNC(int) _PyGen_SetStopIterationValue(PyObject *);
 PyAPI_FUNC(int) _PyGen_FetchStopIterationValue(PyObject **);
 PyAPI_FUNC(void) _PyGen_Finalize(PyObject *self);
+PyAPI_FUNC(PyCodeObject *) PyGen_GetCode(PyGenObject *gen);
```
Comment (ericsnowcurrently) [inline Include/cpython/genobject.h:47]: While I can see the *potential* general utility of this function, does it make sense to add it to the *public* API right now?  (No one has ever asked for it before.  How would someone have gotten this info before?  Through getattr on the the Python object?)  For me it would be enough if you anticipate it would be useful enough and a stable feature, which it essentially is.

If there isn't a solid reason to add this to the public API then I'd expect the function to be "private" (leading underscore) and, maybe, for the name to reflect that it only exists for testing.

**Review by ericsnowcurrently — APPROVED** [review]
mostly LGTM

There's one minor comment, but I'll leave the resolution up to you.

File: Include/cpython/genobject.h (hunk 2 above)
Comment (markshannon) [reply Include/cpython/genobject.h:47]: The problem is that if there is no accessor function, but the field appears to match the Python attribute, C programmers just access the field directly (usually Cython, but there are others).

We can't really complain about it if there is no API function for them to use.

Comment (ericsnowcurrently) [reply Include/cpython/genobject.h:47]: Yeah, makes sense.

**Comment by markshannon** [comment]
I've converted this to a draft because I want to make the following changes first.

* Change the `_PyInterpreterFrame` so that the code object is the first field.
* Add back the `gi_code` field as a union overlapping the frame, ensuring that code accessing `gi_code` will still work, and allowing us to deprecate the field.

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

**Comment by markshannon** [comment]
I'm removing the `gi_code` attribute.
Any C code that accesses the `gi_code` attribute might clear it, resulting in a double free. It is safer to just remove it.
